frame/support/src/traits.rs

```diff
@@ -1185,6 +1185,75 @@ pub trait VestingSchedule<AccountId> {
 	fn remove_vesting_schedule(who: &AccountId);
 }
 
+/// Abstraction over a fungible assets system.
+pub trait Token<Source> {
+	/// The balance of an account.
+	type Balance: AtLeast32BitUnsigned + FullCodec + Copy + MaybeSerializeDeserialize + Debug +
+		Default + HasCompact;
+
+	type AssetId: Member + Parameter + Default + Copy + HasCompact;
+
+	// PUBLIC IMMUTABLES
+	/// The total amount of issuance in the system for a specific asset.
+	fn total_issuance(id: Self::AssetId) -> Self::Balance;
+
+	/// The balance of a given account.
+	///
+	/// This is the only balance that matters in terms of most operations on tokens. It alone
+	/// is used to determine the balance when in the contract execution environment. When this
+	/// balance falls below the value of `ExistentialDeposit`, then the 'current account' is
+	/// deleted: specifically `FreeBalance`.
+	///
+	/// `system::AccountNonce` is also deleted if `ReservedBalance` is also zero (it also gets
+	/// collapsed to zero if it ever becomes less than `ExistentialDeposit`.
+	fn balance_of(id: Self::AssetId, who: Source) -> Self::Balance;
+
+	/// Returns the remaining number of tokens a spender can spend for a specific asset.
+	fn allowance(id: Self::AssetId, owner: Source, spender: Source) -> Self::Balance;
+
+	// PUBLIC MUTABLES (DANGEROUS)
+
+	/// Transfer some liquid free balance of an asset to another account.
+	fn transfer_asset(id: Self::AssetId, origin: Source, dest: Source, amount: Self::Balance) -> DispatchResult;
+
+	/// Approve some amount of an asset to be spent by another account.
+	fn approve(id: Self::AssetId, who: Source, spender: Source, amount: Self::Balance) -> DispatchResult;
```

Comment:
I really dislike the approve / transfer_from from ERC20. It makes sense at that time when ERC20 was proposed but that does mean it is still a good pattern we should follow here.

Reply:
I'll add a feature flag to this as we'll likely try to mimic behavior wrt to EVM.

Reply:
I think another more clean way is move the erc20 feature into another trait, so the ERC20 trait does not need to be feature flag protected but the implementation can be.

Reply:
I agree here. don't use a feature flag.